internal/lib/conmonmon.go

```diff
@@ -0,0 +1,127 @@
+package lib
+
+import (
+	"sync"
+	"time"
+
+	"github.com/cri-o/cri-o/internal/oci"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+var sleepTime = 5 * time.Minute
+
+// conmonmon is a struct responsible for monitoring conmons
+// it contains a map of containers -> conmonPID, and sleeps on
+// a loop, waiting for a conmon to die. if it has, it kills the associated
+// container.
+type conmonmon struct {
+	conmons   map[*oci.Container]int
+	mu        sync.RWMutex
+	closeChan chan bool
+	runtime   *oci.Runtime
+}
+
+// newConmonmon creates a new conmonmon instance
+// given a runtime. It also starts the monitoring routine
+func newConmonmon(r *oci.Runtime) *conmonmon {
+	cmm := conmonmon{
+		conmons: make(map[*oci.Container]int),
+		runtime: r,
+	}
+	go cmm.monitorConmons()
+	return &cmm
+}
+
+// monitorConmons sits on a loop and sleeps.
+// after waking, it signals to the conmons, checking
+// if they're still alive.
+func (c *conmonmon) monitorConmons() {
+	for {
+		select {
+		case <-time.After(sleepTime):
+			c.signalConmons()
+		case <-c.closeChan:
+			return
+		}
+	}
+}
+
+// signalConmons loops through all available conmons and
+// sends a kill 0 to them, checking if they're still alive
+// if they're not, the container is killed and we spoof
+// an OOM event for the container
+func (c *conmonmon) signalConmons() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for ctr, conmonPID := range c.conmons {
+		status := ctr.State().Status
+		if status == oci.ContainerStateRunning {
+			if err := unix.Kill(conmonPID, 0); err != nil {
```

Comment:
5 minutes is a long time, and it might be possible the PID was already reused.

Could we switch the Kill(conmonPID, 0) check to verify instead these conditions:

- `readlink -f /proc/$conmonPID/ns/mnt == readlink -f /proc/self/ns/mnt`
- `readlink -f /proc/$conmonPID/exe` points to the conmon executable
- `$CTR_ID` is present in the `/proc/$conmonPID/cmdline` file

while not perfect, at least we are sure it is a conmon executable related to the container and it is not running in a different mount namespace.

Reply:
We could also record the conmon process start time and match that.

Reply:
we could also record `ps -p $PID -o cmd,pid,lstart` and compare the strings

Reply:
I think we could avoid all of these checks with it (maybe keeping the kill?)

Reply:
I briefly experimented using containers/psgo , I could also give that a whirrl

Reply:
@mrunalp getting the start time would either require https://github.com/containers/psgo/issues/61 or would require us pretty much reimplement psgo and doing that ^ so I'm going to see what I can do over there :) in the meantime, I think psgo is a good way for us to verify it's the process we want, so I switched over to that